libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -775,6 +799,31 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
 {
 	setLocation(_binOp);
 
+	FunctionDefinition const* function = *_binOp.annotation().userDefinedFunction;
+	if (function)
+	{
+		_binOp.leftExpression().accept(*this);
+		_binOp.rightExpression().accept(*this);
+		setLocation(_binOp);
+
+		solAssert(function->isImplemented());
+		solAssert(function->isFree());
+		solAssert(function->parameters().size() == 2);
+		solAssert(function->returnParameters().size() == 1);
+		solAssert(*function->returnParameters()[0]->type() == *_binOp.annotation().type);
+
+		string left = expressionAsType(_binOp.leftExpression(), *function->parameters()[0]->type());
+		string right = expressionAsType(_binOp.rightExpression(), *function->parameters()[1]->type());
```

Comment:
```suggestion
		string left = expressionAsCleanedType(_binOp.leftExpression(), *function->parameters()[0]->type());
		string right = expressionAsCleanedType(_binOp.rightExpression(), *function->parameters()[1]->type());
```

Or why don't we need cleanup here?

Reply:
Not sure. I will need to check. I was assuming that the cleanup happens somewhere inside `expressionAsType()`. It did not break any tests, but to be honest, I don't have tests for cleanup in custom operators (do we even have such tests for built-in ones?) so it could actually be broken.

Reply:
Yeah, this is generally done a bit confusingly, since in the middle of building this originally we switched strategies between assuming stuff being always clean and always cleaning stuff explicitly (which can make a difference in the presence of inline assembly only, but there it can...)...
Anyways, since the comparison operators do this, we should do it here as well. I'd guess the following would yield different results depending on it:

```solidity
type T is uint8;
function f(T x, T y) returns (T z) {
  assembly { z := shr(8, x) }
}
using {f as +} for T;
contract C {
  function test() external returns (T) {
    T a;
    assembly { a := 0x4200 }
    return a + a;
  }
}

```

Probably no way to trigger a difference without inline assembly on both sides :-).

Reply:
After looking at this, I think the current behavior (i.e. no cleanup) is correct because it matches what happens when you call the function directly. In your test, if you replace `a + a` with `f(a, a)`, you get 42, not 0.

Also, I think that the whether we do cleanup for built-in operators or not does not matter here because no one can really observe whether it happens or not. In either case the implementation provided by the compiler will account for it and calculate the result correctly. It may or may not clean the result (I did not check), but even then I think the behavior of user-defined operators should be the same as when the function is called directly and does not necessarily have to match the behavior of the built-in operator.

I added tests (`calling_operator_parameter_cleanup.sol` and `calling_operator_return_parameter_cleanup.sol`) but due to this I put in the expectation that the parameters and return values are **not** cleaned.

Reply:
I mean, you have a test case right there that observes it, so it can be observed :-).
The part with function calls is some argument, but arguably we could even consider that a bug...
It's a bit tricky due to that, but I'd *still* do the cleanup actually... let me think...

Reply:
It's not about matching the behaviour of builtin operators, it's about avoiding weird behaviour - this is especially important since operator definitions are quite likely to be done in assembly.
```
type T is uint8;
function f(T x, T y) returns (bool z) {
  assembly { z := eq(x, y) }
}
using {f as ==} for T;
contract C {
  function test() external returns (bool) {
    T a; T b = T.wrap(0);
    assembly { a := 0x4200 }
    return a == b;
  }
}
```
This *should* return true...

Reply:
But yeah, a bit unfortunate that we don't do it for internal function calls, which makes this inconsistent no matter what...

Reply:
I think it's consistent to do the same thing we do for internal function calls. It is just syntactic sugar after all.

Reply:
if you implement a custom comparison operator in inline assembly, you have to think about what to do with unclean inputs. This way we actually leave it open to the implementer what to do.

Reply:
Well, @chriseth is also for keeping it without cleanup, so let's keep it as is.

Reply:
> if you implement a custom comparison operator in inline assembly, you have to think about what to do with unclean inputs. This way we actually leave it open to the implementer what to do.

Not sure if this isn't a bit reckless... I don't expect people who write implementations for these will take unclean input into account - I mean we missed that a few times ourselves in the past...

But we can hope that small types as underlying types is uncommon anyways and should properly rethink the current cleanup mess in 0.9.

Reply:
> I mean, you have a test case right there that observes it, so it can be observed :-).

This case observes the behavior of a user-defined operator. How can you observe that for a built-in one? You cannot really inject assembly into it.

Reply:
> This should return true...

Why thought? It accesses the whole slots containing `x` and `y` rather than just their `uint8` part. I'm actually rather surprised that we even want to guarantee this level of cleanup.

And, actually, why would it ever be ok to clean up for an operator and not a function? You can do the exact same mistake when calling a function that you implemented in assembly.